### PR TITLE
optimize get shard state from db

### PR DESCRIPTION
**Pull request by xiemylogos**
1、fix get shard config,modify verify shardtx
2、fix sset cross shard msg sig
3、fix get shardtx from parrent ledger

### CODE REVIEW

File: consensus/utils/cross_shard_utils.go [hunk 1]
```diff
@@ -110,3 +113,31 @@ func BuildCrossShardMsgs(signer *account.Account, lgr *ledger.Ledger, blkNum uin
...
 	return builtMsgs, hashRoot, nil
 }
+
+func BuildCrossShardMsgHash(shardMsgs []xshard_types.CommonShardMsg) ([]common.Uint256, common.Uint256) {
+	shardList := make([]common.ShardID, 0)
+	shardMsgMap := make(map[common.ShardID][]xshard_types.CommonShardMsg)
+	for _, msg := range shardMsgs {
+		targetShardID := msg.GetTargetShardID()
+		if _, present := shardMsgMap[targetShardID]; !present {
+			shardList = append(shardList, targetShardID)
```
Comment (qiluge) [inline consensus/utils/cross_shard_utils.go:124]: shardMsgMap[targetShardID] = make([]xshard_types.CommonShardMsg, 0)

File: consensus/vbft/utils.go [hunk 2]
```diff
@@ -356,11 +349,14 @@ func getShardConfigByShardID(lgr *ledger.Ledger, shardID common.ShardID, blkNum
 		return nil, err
 	}
 	var blkHeight uint32
+	sort.Slice(heights, func(i, j int) bool { return heights[i] < heights[j] })
```
Comment (qiluge) [inline consensus/vbft/utils.go]: heights should be a increasing order, optimize these code and AddShardConsensusHeight function

File: core/store/ledgerstore/ledger_store.go [hunk 3]
```diff
@@ -858,12 +859,52 @@ func (this *LedgerStoreImp) saveCrossShardGovernanceData(block *types.Block, sha
...
+		case shardstates.EVENT_SHARD_ACTIVATED:
+			evt := &shardstates.ShardActiveEvent{}
+			if err := evt.Deserialization(common.NewZeroCopySource(shardEvt.Payload)); err != nil {
+				log.Errorf("deserialize shard activation event: %s", err)
+				continue
+			}
+			shardState := &shardstates.ShardState{}
+			shardIDBytes := utils.GetUint64Bytes(evt.ShardID.ToUint64())
+			key := append([]byte(shardmgmt.KEY_SHARD_STATE), shardIDBytes...)
+			data, unkown := vconfig.GetRawStorageItemFromMemDb(overlayDB, utils.ShardMgmtContractAddress, key)
```
Comment (qiluge) [inline core/store/ledgerstore/ledger_store.go]: typos: unknown

File: core/chainmgr/xshard/xshard_block.go [hunk 4]
```diff
@@ -268,12 +280,12 @@ func DelCrossShardTxs(lgr *ledger.Ledger, crossShardTxs map[uint64][]*types.Cros
...
 		for _, shardTx := range shardTxs {
-			if _, present := pool.Shards[shardID]; !present {
+			if msg, present := pool.Shards[shardID]; !present {
 				log.Infof("delcrossshardtxs shardID:%d,not exist", shardID)
 				return nil
 			} else {
 				log.Infof("delcrossshardtxs shardID:%d", shardID)
-				//delete(crossShardTxInfos, shardTx.ShardMsg.CrossShardMsgRoot)
+				delete(msg, shardTx.ShardMsg.PreCrossShardMsgHash)
 				SaveCrossShardHash(lgr, shardID, shardTx.ShardMsg.PreCrossShardMsgHash)
```
Comment (qiluge) [inline core/chainmgr/xshard/xshard_block.go]: handle this error

**Review by qiluge — APPROVED** [review]

**Review by Honglei-Cong — APPROVED** [review]

**Pull request merged**